File: devel/project_files.pl
#! /usr/bin/perl -w
#
# project_files.pl -- find all files that contribute to some ENV
#
# (C) 2010 jw@suse.de, Novell Inc.
# Distribute under GPLv2 or GPLv3
#
# 2010-02-25, jw, v0.1 -- inital draught
# 2010-05-25, jw, v0.2 -- renamed from dm_files to project_files
#                         added support for svn based retry in case of xml parse errors.
# 2010-05-30, jw, v0.3 -- made code reusable, moved into SUSEDOC.pm

use FindBin;
use lib "$FindBin::Bin/../lib/perl/SUSEDOC"; 
use lib "/usr/share/libsusedoc/lib/perl/SUSEDOC";

use SUSEDOC;
use XML::Parser;
use Data::Dumper;
use Getopt::Long;
use Pod::Usage;

my $version = $SUSEDOC::VERSION;
my $verbose = 1;

my ($help, $prof, $profos, $all_files, $img_only, $long_list);

my %opt = ( prev_cmd => undef, restore_cmd => undef, retry_limit => 10 );

## BUG ALERT: -P without parameter, gives this misleading error message:
##  Option p requires an argument
#
GetOptions(
	"all" 	   => \$all_files,
	"verbose+" => \$verbose,
	"quiet"    => sub { $verbose = 0; },
	"images"   => \$img_only,
	"long"     => \$long_list,
	"profos=s" => \$profos,
	"help|?"   => \$help,
	"previous-file-cmd|P=s"  => \$opt{prev_cmd},
	"cleanup-file-cmd|C=s"   => \$opt{restore_cmd},
	"retry-limit=i"	       => \$opt{retry_limit},
);

pod2usage(-verbose => 1, -msg => qq{
project_files V$version Usage: 

$0 [options] startfile.xml
$0 [options] ENV-startfile
env MAIN=MAIN.file.xml $0 [options]

Valid options are:
 -v	Be more verbose. Default: $verbose.
 -q     Be quiet, not verbose.

 -a --all
 	Print all files, xml and images. Default: xml
 -i --images
 	Print images only. Default: xml files.

 -p --profos
 	Profile for one operating system.
	Default from ENV{PROFOS}, otherwise: any.
	Use '-p --' to ignore profiling from ENV.
	Use '-p -osuse' to exclude openSUSE specifc parts.

 -l --long
 	List 3 column layout: filename, profiling, parent_file[node...]
	A list of node numbers, specifies the xml-node where the file was included.
	Default: filenames only.

 -P --previous-file-cmd "svn revert '%s'; svn up -r PREV '%s'"
 	Run the specified command if XML::Parse fails. Useful to e.g. fetch an earlier version
	from svn, hoping to avoid a newly introduced bug.

 -C --cleanup-file-cmd  "svn up -r HEAD '%s'"
 	Run the specifed command, after all error recovery attempts. This should restore the
	file to the state it was before.

 -r --retry-limit NNN
 	Run previous-file-cmd at most NNN times for a file. Default $opt{retry_limit}

 -h -? --help
 	Show this online help.

project_files does a quick file tree walk through docbook xml files.
It follows <xi:include href=...> and finds <imagedata fileref=...>
and optionally honors profiling information while doing so.

}) if $help;

my $start = shift;
my $ctx = new SUSEDOC(env => $start, profos => $profos);

die "please give an ENV or xml file to start with (or source an ENV file).\n" 
  unless $ctx->env->{MAIN};

my $list = $ctx->project_files( %opt );

for my $f (sort keys %$list)
  {
    my $print = $all_files || $list->{$f}{type} eq 'xml';
    $print = $list->{$f}{type} eq 'img' if $img_only;
    next unless $print;

    if ($long_list)
      {
        my $os = $list->{$f}{os};
	my $prof = '-';
	if (keys %$os and !$os->{'*'})
	  {
	    $prof = 'os="'.join('","', sort keys %$os).'"';
	  }
        printf "%-50s %-10s %s\n", $f, $prof, $list->{$f}{via};
      }
    else
      {
        print "$f\n" if $print;
      }
  }

exit 0;
#######################################################

File: devel/project_files.py
#!/usr/bin/python
#
# project_files.py -- find all files that contribute to some ENV
#                     using a library written in perl.
#
# (C) 2010 jw@suse.de, Novell Inc.
# Distribute under GPLv2 or GPLv3
#
# dependencies:
#   home:babelworx:python perlmodule
#   devel:tools:documentation perl-SUSEDOC
#
# TODO: 
#  This is a cheap plastic immitation of project_files.pl -- see there.
#  It has no command line options, and does not format the output nicely.
#  Needs help of a python-literate person.
#  --> toms: Done, but needs improvement if everything is correctly
#            implemented
# 

__version__="$Id: xxx $"

import os,sys,pprint
libdir = os.path.abspath(os.path.dirname(sys.argv[0]) + '/../lib/python')
print libdir
sys.path.insert(0, libdir)


from susedoc import *

import optparse

def parse():
  """Create an OptionParser and analyze the commandline options"""
  # Used descriptions from the Perl script:
  parser = optparse.OptionParser(
     usage="\n  %prog [options] startfile.xml\n" \
           "  %prog [options] ENV-startfile",
     version="%prog " + __version__[3:-1],
     description="Find all files that contribute to some ENV")
  parser.add_option("-v", "--verbose",
      dest="verbose",
      action="store_true",
      help="Be more verbose. Default: %default")
  parser.add_option("-q", "--quiet",
      dest="verbose",
      action="store_false",
      help="Be quiet, not verbose") 	
  # I'm not sure, if this is really intended:
  parser.add_option("-a", "--all",
      dest="all",
      choices=("xml", "files", "images"),
      help="Print all files, xml and images. Default: %default")
  # I'm not sure here either:
  parser.add_option("-i", "--images",
      dest="images",
      # choices=("xml", "files", "images"),
      help="Print images only. Default: xml files (FIXME)")
  parser.add_option("-p", "--profos",
      dest="profos",
      help="Profile for one operating system. " \
           "Default from ENV{PROFOS}, otherwise: any. " \
           "Use '-p --' to ignore profiling from ENV. " \
           "Use '-p -osuse' to exclude openSUSE specifc parts")
  parser.add_option("-l", "--long",
      dest="long",
      help="List 3 column layout: filename, profiling,parent_file[node...] "\
           "A list of node numbers, specifies the xml-node where " \
           "the file was included. " \
           "Default: filenames only.")
  parser.add_option("-P", "--previous-file-cmd",
      dest="previous_file_cmd",
      help="Run the specified command if XML::Parse fails. " \
            "Useful to e.g. fetch an earlier version from svn, "\
            "hoping to avoid a newly introduced bug.")
  parser.add_option("-C", "--cleanup-file-cmd",
      dest="cleanup_file_cmd",
      help="Run the specifed command, after all error recovery attempts. " \
           "This should restore the	file to the state it was before.")
  parser.add_option("-r", "--retry-limit",
      dest="retry_limit",
      type="int",
      help="Run previous-file-cmd at most NNN times for a file. " \
            "Default %default") 
  # Set all the default values for the previous options:
  parser.set_defaults(verbose=True,
                      all="xml",
                      )
  (options, args) = parser.parse_args()
  # Just for debugging purpose, can be removed later:
  print options, args
  return options, args
  # sys.exit(10)
  

if __name__=="__main__":
  # Populate the options and arguments.
  # options is a list, args is a dictionary: 
  options, args = parse()
  envfile = options[0]
  
  ctx = SUSEDOC(env=envfile)
  # ctx = SUSEDOC(**{ 'env': envfile })

  list = ctx.project_files()

  pprint.pprint(copy_in(list))

  for f in sorted(list.keys()):
    if (list[f]['type'] == 'xml'):
      print f

